lib/Injections/NetworkAgent.js

```diff
@@ -0,0 +1,306 @@
+      type: 'XHR',
+      response: {
+        url: getRequestUrl(request),
+        status: response.statusCode,
+        statusText: response.statusMessage,
+        headers: getCorrectHeaders(response),
+        mimeType: getMimeType(response),
+        connectionReused: request.shouldKeepAlive,
+        connectionId: request.connection.__inspector_ID__,
+        encodedDataLength: -1,
+        fromDiskCache: false,
+        fromServiceWorker: false,
+        timing: request.__inspector_timing__.json,
+        headersText: restoreHeadersText(response),
+        requestHeaders: renderHeaders(request),
+        requestHeadersText: request._header
+      }
+    };
+  }
+
+  debug.registerEvent('Network.requestWillBeSent');
+  debug.registerEvent('Network.responseReceived');
+  debug.registerEvent('Network._responseReceived');
+  debug.registerEvent('Network.dataReceived');
+  debug.registerEvent('Network._dataReceived');
+  debug.registerEvent('Network.loadingFinished');
+  debug.registerEvent('Network._loadingFinished');
+
+  wrapHttpRequests();
+  debug.once('close', unwrapHttpRequests);
+
+  function wrapHttpRequests() {
+    var lastRequestId = 0,
+        lastConnectionId = 0;
+
+    function handleHttpRequest() {
+      var timing = new Timing();
+
+      this.__inspector_timing__ = timing;
+      this.__inspector_ID__ = '' + lastRequestId++;
+
+      var requestInfo = constructRequestInfo(this);
+
+      this.once('socket', handleSocket.bind(this, requestInfo));
+      this.once('response', handleHttpResponse);
+      handleRequestData.call(this, requestInfo);
+
+      deferToConnect.apply(this, arguments);
+      timing.dnsStart();
+    }
+
+    function handleRequestData(requestInfo) {
+      var oldWrite = this.write;
+
+      this.write = function(chunk) {
+        requestInfo.request.postData += chunk || '';
+        return oldWrite.apply(this, arguments);
+      };
+    }
+
+    function handleSocket(requestInfo, socket) {
+      socket.__inspector_ID__ = socket.__inspector_ID__ || '' + lastConnectionId++;
+
+      this.__inspector_timing__.dnsEnd().connectStart();
+
+      socket.once('connect', function() {
+        this.__inspector_timing__.connectEnd().sendStart();
+        setImmediate(this.__inspector_timing__.sendEnd.bind(this.__inspector_timing__));
+      }.bind(this));
+
+      requestInfo.request.url = requestInfo.documentURL = getRequestUrl(this);
+      debug.emitEvent('Network.requestWillBeSent', requestInfo);
+    }
+
+    function handleHttpResponse(response) {
+      var request = response.req;
+      request.__inspector_timing__.receiveHeadersEnd();
+
+      var requestId = request.__inspector_ID__,
+        responseInfo = constructResponseInfo(response),
+        dataLength = 0;
+
+      debug.emitEvent('Network.responseReceived', responseInfo);
+      debug.emitEvent('Network._responseReceived', {
+        requestId: requestId
+      });
+
+      response.on('data', function(chunk) {
```

Comment:
To be honest, I don't know enough details about stream to tell what are the (possible) side effect of piping the response to another stream, but my gut feeling is that such solution is still not bullet proof.

I think you'll have to dig deeper and hook into an internal API. Below is a snippet of code that hooks into `res.push` and works for me in Node v0.10 and io.js v2.2.1:

```js
var http = require('http');

http.get('http://www.root.cz', function(res) {
  var push = res.push;
  res.push = function(chunk) {
    if (chunk) {
      // handle intercepted chunk of response data
      process.stderr.write(chunk.toString('utf-8'));
    }
    // call the original push method to preserve the observed behaviour
    return push.apply(this, arguments);
  };
  res.pipe(process.stdout);
});
```

The script prints the original response stream to stdout and the intercepted response stream to stderr. I ran it with `
node script.js 1>1.log 2>2.log` and called `diff -u 1.log 2.log` to verify that the output is the same.